File: src/commenteme/collectors/gpu.py
"""Leitura da GPU em Apple Silicon, SEM sudo, via `ioreg`.

No M-series não existe `nvidia-smi`. Os dados de utilização e memória da GPU
ficam expostos no IORegistry, na classe IOAccelerator (ex.: AGXAcceleratorG16X),
dentro de "PerformanceStatistics". Lemos com `ioreg` e extraímos por regex.
"""

from __future__ import annotations

import re
import subprocess

# Cada chave aparece no dump como:  "Device Utilization %"=9
_PATTERNS = {
    "util_pct": "Device Utilization %",
    "mem_used_bytes": "In use system memory",
    "mem_alloc_bytes": "Alloc system memory",
}


def _find_int(text: str, key: str):
    # ancora no '"key"=' exato para não casar com "In use system memory (driver)"
    m = re.search(r'"%s"\s*=\s*(\d+)' % re.escape(key), text)
    return int(m.group(1)) if m else None


def read_gpu() -> dict:
    """Retorna {util_pct, mem_used_bytes, mem_alloc_bytes}. Valores None se ausentes."""
    try:
        out = subprocess.run(
            ["ioreg", "-r", "-d", "1", "-w", "0", "-c", "IOAccelerator"],
            capture_output=True,
            text=True,
            timeout=2,
        ).stdout
    except Exception:
        return {k: None for k in _PATTERNS}

    return {field: _find_int(out, key) for field, key in _PATTERNS.items()}


if __name__ == "__main__":  # debug rápido: python -m commenteme.collectors.gpu
    print(read_gpu())

File: src/collectors/gpu.ts
/**
 * Apple Silicon GPU via `ioreg` (no sudo). Parses utilization and system
 * memory in use / allocated from the IOAccelerator registry entry.
 */

import { execa } from "execa";
import type { GpuData } from "./types.js";

const EMPTY: GpuData = { utilPct: null, memUsedBytes: null, memAllocBytes: null };

export async function readGpu(): Promise<GpuData> {
  try {
    const { stdout } = await execa("ioreg", ["-r", "-d", "1", "-w", "0", "-c", "IOAccelerator"], {
      timeout: 2000,
    });
    const get = (key: string): number | null => {
      const m = stdout.match(new RegExp(`"${escapeRe(key)}"\\s*=\\s*(\\d+)`));
      return m ? Number(m[1]) : null;
    };
    return {
      utilPct: get("Device Utilization %"),
      memUsedBytes: get("In use system memory"),
      memAllocBytes: get("Alloc system memory"),
    };
  } catch {
    return { ...EMPTY };
  }
}

function escapeRe(s: string): string {
  return s.replace(/[.*+?^${}()|[\]\\]/g, "\\$&");
}
